sale_order_invoice_amount/hooks.py

```diff
@@ -0,0 +1,56 @@
+def pre_init_hook(cr):
+    cr.execute(
+        """
+ALTER TABLE
+    sale_order
+ADD COLUMN
+    invoiced_amount numeric,
+ADD COLUMN
+    uninvoiced_amount numeric
+        """
+    )
+    cr.execute(
+        """
+UPDATE
+    sale_order
+SET
+    invoiced_amount = 0.0,
+    uninvoiced_amount = sale_order.amount_total
+WHERE
+    sale_order.state = 'cancel'
+    """
+    )
+    cr.execute(
+        """
+WITH amt AS(
+    SELECT
```

Comment:
Tested cases:

1. Sale Order with delivered quantities but no invoices
2. Sale Order partially invoiced
3. Sale Order totally invoiced

```sql
 sale_order_id | invoiced_amount | uninvoiced_amount | state 
---------------+-----------------+-------------------+-------
            26 |          339.25 |           3053.25 | sale
            27 |         3392.50 |               0.0 | sale
(2 rows)

 id | amount_total | state 
----+--------------+-------
 21 |      3392.50 | sale
(1 row)
```

The first case is not in `amt` selection, so `uninvoiced_amount` is set with `NULL` instead the correct value.

Try changing the condition for `account_move` with the following:

```sql
AND (
            account_move IS NULL
            OR (
                account_move.move_type IN ('out_invoice', 'out_refund')
                AND account_move.state != 'cancel'
            )
    )
```

Result:

```sql
 sale_order_id | invoiced_amount | uninvoiced_amount | sale_order_state 
---------------+-----------------+-------------------+------------------
            26 |          339.25 |           3053.25 | sale
            27 |         3392.50 |               0.0 | sale
            21 |                 |               0.0 | sale
(3 rows)
```

Now you have to change how `invoiced_amount` and `uninvoiced_amount` is calculated for that case.

Try changing the selection with the following:

```sql
SELECT
    sale_order.id AS sale_order_id,
    COALESCE(SUM(account_move.amount_total_signed), 0) AS invoiced_amount,
    CASE
        WHEN SUM(account_move.amount_total_signed) IS NULL THEN sale_order.amount_total
        WHEN sale_order.amount_total - SUM(
            account_move.amount_total_signed) > 0.0 THEN sale_order.amount_total - SUM(
            account_move.amount_total_signed)
        ELSE 0.0
    END AS uninvoiced_amount
```

Result:

```sql
 sale_order_id | invoiced_amount | uninvoiced_amount | sale_order_state 
---------------+-----------------+-------------------+------------------
            26 |          339.25 |           3053.25 | sale
            27 |         3392.50 |               0.0 | sale
            21 |               0 |           3392.50 | sale
(3 rows)
```